query/src/client/QueryMetadataEditor/QueryMetadataEditor.tsx

```diff
@@ -0,0 +1,396 @@
+    showViewDataConfirmationModal: boolean,
+    navigateAfterSave: boolean
+}
+
+export class App extends PureComponent<any, Partial<IAppState>> {
+
+    constructor(props) {
+        super(props);
+
+        const { schemaName, queryName, returnUrl } = ActionURL.getParameters();
+
+        let messages = List<IBannerMessage>();
+        if ((!schemaName || !queryName)) {
+            let msg =  'Missing required parameter: schemaName and queryName.';
+            let msgType = 'danger';
+            let bannerMsg ={message : msg, messageType : msgType};
+            messages = messages.push(bannerMsg);
+        }
+
+        this.state = {
+            schemaName,
+            queryName,
+            returnUrl,
+            messages,
+            showAlias : false,
+            showSave: true,
+            dirty: false,
+            showEditSourceConfirmationModal: false,
+            showResetConfirmationModal: false,
+            showViewDataConfirmationModal: false,
+            navigateAfterSave: false
+        };
+    }
+
+    componentDidMount() {
+        const { schemaName, queryName, messages } = this.state;
+
+        if (schemaName && queryName) {
+            fetchQueryMetadata(schemaName, queryName)
+                .then((domain) => this.setState(() => ({domain})))
+                .catch((error) => {
+                    this.setState(() => ({
+                        messages: messages.set(0, {message: error.exception, messageType: 'danger'})
+                    }));
+                });
+        }
+        window.addEventListener("beforeunload", this.handleWindowBeforeUnload);
+    }
+
+    handleWindowBeforeUnload = (event) => {
+        if (this.state.dirty) {
+            event.returnValue = 'Changes you made may not be saved.';
+        }
+    };
+
+    onChangeHandler = (newDomain, dirty) => {
+        this.setState((state) => ({
+            domain: newDomain,
+            dirty: state.dirty || dirty, // if the state is already dirty, leave it as such
+            showSave: true
+        }));
+    };
+
+    showMessage = (message: string, messageType: string, index: number, additionalState?: Partial<IAppState>) => {
+        const { messages } = this.state;
+
+        this.setState(Object.assign({}, additionalState, {
+            messages : messages.set(index, {message: message, messageType: messageType})
+        }));
+    };
+
+    dismissChangeConfirmation = () => {
+        const {showViewDataConfirmationModal, showEditSourceConfirmationModal} = this.state;
+
+        this.setState(() => ({
+            showResetConfirmationModal: false,
+            showViewDataConfirmationModal: false,
+            showEditSourceConfirmationModal: false,
+            dirty: false
+        }));
+
+        if (showViewDataConfirmationModal) {
+            this.navigateToViewData();
+        }
+
+        if (showEditSourceConfirmationModal) {
+            this.navigateToEditSource();
+        }
```

Comment:
To ensure that these happen after the state change is complete, you probably want to put them into the callback for the setState function. So it would look something like this:

```suggestion
        this.setState(() => ({
            showResetConfirmationModal: false,
            showViewDataConfirmationModal: false,
            showEditSourceConfirmationModal: false,
            dirty: false
        }), () => {
                if (showViewDataConfirmationModal) {
                   this.navigateToViewData();
               }

               if (showEditSourceConfirmationModal) {
                   this.navigateToEditSource();
               }
        });
```